### PR TITLE
[Fix](jdbc catalog) Pass conjuncts to JdbcScanNode and FileScanNode before doing finalize.

### PR DESCRIPTION
**Pull request by Jibing-Li**
JdbcScanNode need to use the conjuncts to generate sql in finalize function. But the conjuncts have not passed to JdbcScanNode yet while calling finalize. This pr is to pass the conjuncts to scan node before using it to avoid scan the whole table.

```
mysql> explain select * from jdbc.doris_test.ex_tb0 where id=111;
+-----------------------------------------------------------------------------------+
| Explain String                                                                    |
+-----------------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                                   |
|   OUTPUT EXPRS:                                                                   |
|     id[#0]                                                                        |
|     name[#1]                                                                      |
|   PARTITION: UNPARTITIONED                                                        |
|                                                                                   |
|   VRESULT SINK                                                                    |
|                                                                                   |
|   1:VEXCHANGE                                                                     |
|      offset: 0                                                                    |
|                                                                                   |
| PLAN FRAGMENT 1                                                                   |
|                                                                                   |
|   PARTITION: RANDOM                                                               |
|                                                                                   |
|   STREAM DATA SINK                                                                |
|     EXCHANGE ID: 01                                                               |
|     UNPARTITIONED                                                                 |
|                                                                                   |
|   0:VJdbcScanNode                                                                 |
|      TABLE: `doris_test`.`ex_tb0`                                                 |
|      QUERY: SELECT `id`, `name` FROM `doris_test`.`ex_tb0` WHERE (`id`[#0] = 111) |
+-----------------------------------------------------------------------------------+
```

### CODE REVIEW

**Review by morningman — COMMENTED** [review]
Please add test case for this.
You can simply use jdbc catalog to connect doris it self in your test cases

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalScanNode.java [hunk 1]
```diff
@@ -42,6 +46,8 @@ public abstract class ExternalScanNode extends ScanNode {
 
     // set to false means this scan node does not need to check column priv.
     protected boolean needCheckColumnPriv;
+    private Set<Expression> nereidsConjuncts;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalScanNode.java:49]: nereidsConjuncts is a bad name, please rename it to something more meaningful

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalScanNode.java [hunk 2]
```diff
@@ -42,6 +46,8 @@ public abstract class ExternalScanNode extends ScanNode {
...
+    private PlanTranslatorContext context;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalScanNode.java:50]: scannode should not have a PlanTranslatorContext attr

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalScanNode.java [hunk 3]
```diff
@@ -90,4 +97,20 @@ public boolean needToCheckColumnPriv() {
...
+    public void setContext(PlanTranslatorContext context) {
+        this.context = context;
+    }
+
+    private void translateConjuncts() {
+        if (nereidsConjuncts != null) {
+            nereidsConjuncts.stream()
+                .map(e -> ExpressionTranslator.translate(e, context))
+                    .forEach(conjuncts::add);
+        }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalScanNode.java:114]: u should tranlate in PhysicalPlanTranslator and set legacy planner's expr（List<Expr>）to this node

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/TranslateJdbcConjuncts.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleType;
+import org.apache.doris.nereids.trees.plans.logical.LogicalFilter;
+import org.apache.doris.nereids.trees.plans.logical.LogicalJdbcScan;
+
+/**
+ * Rewrite jdbc plan to set the conjuncts.
+ */
+public class TranslateJdbcConjuncts extends OneRewriteRuleFactory {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/TranslateJdbcConjuncts.java:28]: rename to `PushConjunctsIntoJdbcScan`

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalScanNode.java (hunk 1 above)
Comment (Jibing-Li) [reply fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalScanNode.java]: Removed this attr.

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalScanNode.java (hunk 2 above)
Comment (Jibing-Li) [reply fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalScanNode.java]: removed

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalScanNode.java (hunk 3 above)
Comment (Jibing-Li) [reply fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalScanNode.java]: Moved the translate code to PhysicalPlanTranslator

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/TranslateJdbcConjuncts.java (hunk 4 above)
Comment (Jibing-Li) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/TranslateJdbcConjuncts.java]: Renamed.

**Comment by morrySnow** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalJdbcScan.java [hunk 5]
```diff
@@ -22,34 +22,42 @@
...
      * Constructor for LogicalJdbcScan.
      */
     public LogicalJdbcScan(RelationId id, TableIf table, List<String> qualifier,
                            Optional<GroupExpression> groupExpression,
-                           Optional<LogicalProperties> logicalProperties) {
+                           Optional<LogicalProperties> logicalProperties,
+                           Set<Expression> conjuncts) {
         super(id, PlanType.LOGICAL_JDBC_SCAN, table, qualifier,
                 groupExpression, logicalProperties);
+        this.conjuncts = conjuncts;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalJdbcScan.java:56]:
```suggestion
        this.conjuncts = ImmutableSet.copyOf(Objects.requireNonNull(conjuncts, "conjuncts should not be null"));
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalJdbcScan.java [hunk 6]
```diff
@@ -22,34 +22,42 @@
...
     }
 
     public LogicalJdbcScan(RelationId id, TableIf table, List<String> qualifier) {
-        this(id, table, qualifier, Optional.empty(), Optional.empty());
+        this(id, table, qualifier, Optional.empty(), Optional.empty(), Sets.newHashSet());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalJdbcScan.java:60]:
```suggestion
        this(id, table, qualifier, Optional.empty(), Optional.empty(), ImmutableSet.of());
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalJdbcScan.java [hunk 7]
```diff
@@ -30,28 +31,34 @@
...
+    private final Set<Expression> conjuncts;
+
     /**
      * Constructor for PhysicalJdbcScan.
      */
     public PhysicalJdbcScan(RelationId id, TableIf table, List<String> qualifier,
-            Optional<GroupExpression> groupExpression, LogicalProperties logicalProperties) {
+            Optional<GroupExpression> groupExpression, LogicalProperties logicalProperties, Set<Expression> conjuncts) {
         super(id, PlanType.PHYSICAL_JDBC_SCAN, table, qualifier, groupExpression, logicalProperties);
+        this.conjuncts = conjuncts;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalJdbcScan.java:49]:
```suggestion
        this.conjuncts = ImmutableSet.copyOf(Objects.requireNonNull(conjuncts, "conjuncts should not be null"));;
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalJdbcScan.java [hunk 8]
```diff
@@ -30,28 +31,34 @@
...
      * Constructor for PhysicalJdbcScan.
      */
     public PhysicalJdbcScan(RelationId id, TableIf table, List<String> qualifier,
             Optional<GroupExpression> groupExpression,
-            LogicalProperties logicalProperties, PhysicalProperties physicalProperties, Statistics statistics) {
+            LogicalProperties logicalProperties, PhysicalProperties physicalProperties, Statistics statistics,
+            Set<Expression> conjuncts) {
         super(id, PlanType.PHYSICAL_JDBC_SCAN, table, qualifier, groupExpression,
                 logicalProperties, physicalProperties, statistics);
+        this.conjuncts = conjuncts;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalJdbcScan.java:61]: this.conjuncts = ImmutableSet.copyOf(Objects.requireNonNull(conjuncts, "conjuncts should not be null"));

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.65 seconds
 stream load tsv:          509 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17164657482 Bytes

**Comment by Jibing-Li** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.2 seconds
 stream load tsv:          509 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17169964062 Bytes

**Review by morningman — APPROVED** [review]

**Review by zy-kkk — APPROVED** [review]

**Pull request merged**